src/silo/preprocessing/preprocessor.test.cpp

```diff
@@ -214,67 +552,134 @@ const Scenario PREVENT_LATE_AUTO_CASTING = {
+   }
+   file.close();
+
+   Preprocessor preprocessor(
+      config_with_input_dir,
+      scenario.database_config,
+      scenario.reference_genomes,
+      scenario.lineage_tree
    );
    auto database = preprocessor.preprocess();
 
    const auto database_info = database.getDatabaseInfo();
 
-   EXPECT_EQ(database_info.sequence_count, scenario.expected_sequence_count);
+   EXPECT_EQ(database_info.sequence_count, scenario.assertion.expected_sequence_count);
 
    const silo::query_engine::QueryEngine query_engine(database);
-   const auto result = query_engine.executeQuery(scenario.query);
+   const auto result = query_engine.executeQuery(scenario.assertion.query);
 
    const auto actual = nlohmann::json(result.entries());
-   ASSERT_EQ(actual, scenario.expected_query_result);
+   ASSERT_EQ(actual, scenario.assertion.expected_query_result);
+
+   std::filesystem::remove_all(input_directory);
+}
+
+class InvalidPreprocessorTestFixture : public ::testing::TestWithParam<Scenario<Error>> {};
+
+const Scenario<Error> DUPLICATE_PRIMARY_KEY{
+   .test_name = "duplicate_primary_key",
+   .input_data =
+      []() {
+         std::vector<NdjsonInputLine> result;
+         result.emplace_back(NdjsonInputLine{.metadata = {{"primaryKey", "id_1"}}});
+         result.emplace_back(NdjsonInputLine{.metadata = {{"primaryKey", "id_1"}}});
+         result.emplace_back(NdjsonInputLine{.metadata = {{"primaryKey", "id_2"}}});
+         result.emplace_back(NdjsonInputLine{.metadata = {{"primaryKey", "id_2"}}});
+         return result;
+      },
+   .database_config =
+      {.schema =
+          {.instance_name = "dummy name",
+           .metadata = {{.name = "primaryKey", .type = ValueType::STRING}},
+           .primary_key = "primaryKey"}},
+   .reference_genomes = {},
+   .assertion = {.error_message = "Found 2 duplicate primary key(s): id_1, id_2"}
+};
+
+INSTANTIATE_TEST_SUITE_P(PreprocessorTest, InvalidPreprocessorTestFixture, ::testing::Values(DUPLICATE_PRIMARY_KEY), printTestName<Error>);
+
+TEST_P(InvalidPreprocessorTestFixture, shouldNotProcessData) {
+   const auto scenario = GetParam();
+   auto now = std::chrono::system_clock::now();
+   auto duration = now.time_since_epoch();
+   auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
+   const std::filesystem::path input_directory = fmt::format("test{}", millis);
+   std::filesystem::create_directories(input_directory);
+
+   auto config_with_input_dir = PreprocessingConfig::withDefaults();
+   config_with_input_dir.input_directory = input_directory;
+   config_with_input_dir.intermediate_results_directory = input_directory;
+   config_with_input_dir.ndjson_input_filename = "input.json";
+   config_with_input_dir.validate();
+
+   std::ofstream file(config_with_input_dir.getNdjsonInputFilename().value());
+
+   if (!file.is_open()) {
+      std::cerr << "Could not open file for writing" << std::endl;
+      return;
```

Comment:
Does this return actually make the test silently fail? It prints something to stderr, but the test will probably still be reported as successful?